middleware/auth-utils/grant-manager.js

```diff
@@ -420,14 +420,20 @@ const fetch = (manager, handler, options, params) => {
     options.headers['Content-Length'] = data.length;
 
     const req = getProtocol(options).request(options, (response) => {
-      if (response.statusCode < 200 || response.statusCode > 299) {
-        return reject(response.statusCode + ':' + http.STATUS_CODES[ response.statusCode ]);
-      }
       let json = '';
       response.on('data', (d) => (json += d.toString()));
-      response.on('end', () => {
-        handler(resolve, reject, json);
-      });
+      if (response.statusCode < 200 || response.statusCode > 299) {
+        response.on('end', () => {
+          const grantType = typeof params === 'object' ? (params.grant_type || '') : '';
+          const req = `${options.method}:${options.path} ${grantType}`;
+          const resp = `${response.statusCode}: ${http.STATUS_CODES[response.statusCode]} ${json}`;
+          reject(new Error(resp + ' on ' + req));
+        });
+      } else {
+        response.on('end', () => {
+          handler(resolve, reject, json);
+        });
+      }
```

Comment:
why we couldn't  have one shared handler for `end` action?
```
response.on('end', () => {
    if (response.statusCode < 200 || response.statusCode > 299) {
        const grantType = typeof params === 'object' ? (params.grant_type || '') : '';
        const req = `${options.method}:${options.path} ${grantType}`;
        const resp = `${response.statusCode}: ${http.STATUS_CODES[response.statusCode]} ${json}`;
        reject(new Error(resp + ' on ' + req));
    } else {
        handler(resolve, reject, json);
    }
})
```
?

Reply:
we can have. we will have :)